src/box/txn.c

```diff
@@ -1069,65 +1069,41 @@ txn_add_limbo_entry(struct txn *txn, const struct journal_entry *req)
 	return 0;
 }
 
-int
-txn_commit_try_async(struct txn *txn)
-{
-	struct journal_entry *req;
-
-	ERROR_INJECT(ERRINJ_TXN_COMMIT_ASYNC, {
-		diag_set(ClientError, ER_INJECTION,
-			 "txn commit async injection");
-		goto rollback;
-	});
-
-	if (txn_prepare(txn) != 0)
-		goto rollback;
-
-	if (txn_commit_nop(txn))
-		return 0;
-
-	req = txn_journal_entry_new(txn);
-	if (req == NULL)
-		goto rollback;
-
-	if (txn_has_flag(txn, TXN_WAIT_SYNC) &&
-	    txn_add_limbo_entry(txn, req) != 0) {
-		goto rollback;
-	}
-
-	fiber_set_txn(fiber(), NULL);
-	if (journal_write_try_async(req) != 0) {
-		fiber_set_txn(fiber(), txn);
-		diag_log();
-		goto rollback;
-	}
-
-	return 0;
-
-rollback:
-	assert(txn->fiber == NULL);
-	txn_abort(txn);
-	return -1;
-}
-
-int
-txn_commit(struct txn *txn)
+static int
+txn_commit_impl(struct txn *txn, enum txn_commit_wait_mode wait_mode)
 {
 	struct journal_entry *req;
-
 	txn->fiber = fiber();
-
 	if (txn_prepare(txn) != 0)
 		goto rollback_abort;
-
 	if (txn_commit_nop(txn)) {
 		txn_free(txn);
 		return 0;
 	}
-
 	req = txn_journal_entry_new(txn);
 	if (req == NULL)
 		goto rollback_abort;
+	if (wait_mode != TXN_COMMIT_WAIT_MODE_COMPLETE) {
+		assert(wait_mode == TXN_COMMIT_WAIT_MODE_SUBMIT);
+		ERROR_INJECT(ERRINJ_TXN_COMMIT_ASYNC, {
+			diag_set(ClientError, ER_INJECTION,
+				 "txn commit async injection");
+			goto rollback_abort;
+		});
+		/*
+		 * The main reason why it is disabled is that the sync txn's
+		 * original instance wants to write CONFIRM. But it can't be
+		 * done asynchronously in the current code design. Also it is
+		 * logically strange to commit a synchronous transaction in an
+		 * async way, but the CONFIRM problem is the technical one.
+		 */
+		bool is_original = req->rows[0]->replica_id == 0;
```

Comment:
It seems `txn->n_new_rows > txn->n_local_rows` check would suit even better here.

Reply:
Hm? Why? `txn_update_row_counts()` does the same check for replica ID. It seems `replica_id == 0` is the primary check from which others like row count can be derived.

Reply:
> and IMO it's better to have a single place where we derive some conclusions out of replica id being zero

I totally agree. I would say even more - it would be better to have a function like `txn_is_replayed(txn)`. The problem here is that we don't have even just 2 places which use this "property" in the same way.
- `txn_add_limbo_entry()` uses `req->rows[0]->replica_id` as the origin ID. Transaction originator. It needs the specific ID of the txn author.
- `txn_commit_impl()` now uses `req->rows[0]->replica_id != 0` as a fact that the txn is replayed. It doesn't need to know the ID.
- `txn_update_row_counts()` works with statement, not txn. It looks at `stmt->row->replica_id == 0`. Again as a property whether it is replayed, but not the txn. Just this row.
- `txn_journal_entry_new()` uses `stmt->row->replica_id == 0` to see if the row is created locally or came remotely (replayed != remote).

As you can see, even if I introduce a function like `txn_is_replayed()`, it is going to be used in a single place. If I introduce a function `txn_get_origin_id()`, then will also be used in a single place, and will be lying (return 0) for the locally created but not yet committed txns, and won't tell if it is replayed or not (origin can be us, but can still be replayed during recovery).

Also it is clear, that *in case someday it changes* (the fact that `replica_id != 0` means a replayed txn), everything will break quite spectacularly.

> your check also relies on the fact that when a transaction is remote, local rows from before/on_replace triggers are never put first

Yes. We have that guarantee AFAIU. `txn_journal_entry_new()` builds the journal entry in this way. Non-local rows go first. Moreover, I treat it as a property of "replayed" transactions. Not only remote, but also replayed from WAL during recovery. Otherwise if that isn't true, the stuff above will also break and we will need to reconsider from scratch how do we determine origin of the transactions. Limbo would break for sure.

And you are absolutely correct about me mentioning shakiness of this construction. I would much more like if we would store author ID somewhere explicitly and not deduct it from the rows. But that would be a too serious change, to introduce any sort of transaction headers.

Reply:
Thanks for taking the time to make this overview!

Ok, let's leave the check as is for now.